ProcessLib/HeatConduction/HeatConductionFEM-impl.h

```diff
@@ -0,0 +1,155 @@
+
+    unsigned const n_integration_points =
+        _integration_method.getNumberOfPoints();
+
+    const double porosity_variable = 0.;
+    for (unsigned ip = 0; ip < n_integration_points; ip++)
+    {
+        pos.setIntegrationPoint(ip);
+        auto const& sm = _shape_matrices[ip];
+        auto const& wp = _integration_method.getWeightedPoint(ip);
+        double p = 0.;
+        NumLib::shapeFunctionInterpolate(local_p, sm.N, p);
+        double T = 0.;
+        NumLib::shapeFunctionInterpolate(local_x, sm.N, T);
+
+        // Material parameters of solid phase
+        auto const k_s = _process_data.thermal_conductivity(t, pos)[0];
+        auto const cp_s = _process_data.heat_capacity(t, pos)[0];
+        auto const rho_s = _process_data.density(t, pos)[0];
+
+        // Material parameters of fluid phase
+        double const cp_f = liquid_flow_prop.getHeatCapacity(p, T);
+        double const k_f = liquid_flow_prop.getThermalConductivity(p, T);
+        double const rho_f = liquid_flow_prop.getLiquidDensity(p, T);
+
+        // Material parameter of porosity
+        double const poro =
+            liquid_flow_prop.getPorosity(material_id, porosity_variable, T);
+
+        double const effective_cp =
+            (1.0 - poro) * cp_s * rho_s + poro * cp_f * rho_f;
+        double const effective_K = (1.0 - poro) * k_s + poro * k_f;
+
+        double const integration_factor =
+            sm.detJ * wp.getWeight() * sm.integralMeasure;
+
+        local_M.noalias() +=
+            effective_cp * sm.N.transpose() * sm.N * integration_factor;
+        local_K.noalias() +=
+            sm.dNdx.transpose() * effective_K * sm.dNdx * integration_factor;
+
+        // Compute fluid flow velocity
+        double const mu = liquid_flow_prop.getViscosity(p, T);  // Viscosity
+        GlobalDimVectorType const& darcy_velocity =
+            LiquidFlowVelocityCalculator::calculateVelocity(
+                local_p_vec, sm, perm, mu, rho_f * gravitational_acceleration,
+                gravitational_axis_id);
+        // Add the advection term
+        local_K.noalias() += cp_f * rho_f * sm.N.transpose() *
+                             (darcy_velocity.transpose() * sm.dNdx) *
+                             integration_factor;
+    }
+}
+
+template <typename ShapeFunction, typename IntegrationMethod,
+          unsigned GlobalDim>
+void LocalAssemblerData<ShapeFunction, IntegrationMethod, GlobalDim>::
+    coupling_assemble(double const t, std::vector<double> const& local_x,
+                      std::vector<double>& local_M_data,
+                      std::vector<double>& local_K_data,
+                      std::vector<double>& /*local_b_data*/,
+                      LocalCouplingTerm const& coupled_term)
+{
+    for (auto const& coupled_process_map : coupled_term.coupled_processes)
+    {
+        if (coupled_process_map.first ==
+            std::type_index(typeid(ProcessLib::LiquidFlow::LiquidFlowProcess)))
```

Comment:
I would like to keep this in case there will be other coupled processes to be added on.